File: examples/governance_static_go/domain/compliance_review.go
package domain

import (
	"time"
	"github.com/google/uuid"
	"fmt"
)

type ComplianceReview struct {
	ID string `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	ModelId string `json:"model_id"`
	PolicyId string `json:"policy_id"`
	ReviewerId string `json:"reviewer_id"`
	Outcome string `json:"outcome"`
	Notes string `json:"notes"`
	CompletedAt time.Time `json:"completed_at"`
	Conditions []ReviewCondition `json:"conditions"`
	Status string `json:"status"`
}

func NewComplianceReview(modelId string, policyId string, reviewerId string, outcome string, notes string, completedAt time.Time, conditions []ReviewCondition, status string) *ComplianceReview {
	a := &ComplianceReview{
		ID:        uuid.New().String(),
		ModelId: modelId,
		PolicyId: policyId,
		ReviewerId: reviewerId,
		Outcome: outcome,
		Notes: notes,
		CompletedAt: completedAt,
		Conditions: conditions,
		Status: status,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	return a
}

func (a *ComplianceReview) Validate() error {
	if a.ModelId == "" {
		return &ValidationError{Field: "model_id", Message: "model_id can't be blank"}
	}
	if a.ReviewerId == "" {
		return &ValidationError{Field: "reviewer_id", Message: "reviewer_id can't be blank"}
	}
	if a.Outcome != "" {
		validOutcome := map[string]bool{"approved": true, "rejected": true}
		if !validOutcome[a.Outcome] {
			return &ValidationError{Field: "outcome", Message: fmt.Sprintf("outcome must be one of: approved, rejected, got: %s", a.Outcome)}
		}
	}
	return nil
}

File: examples/governance/compliance_domain/lib/compliance_domain/compliance_review/compliance_review.rb
require 'hecks/mixins/model'

module ComplianceDomain
  class ComplianceReview
    autoload :ReviewCondition, "compliance_domain/compliance_review/review_condition"
    autoload :Lifecycle, "compliance_domain/compliance_review/lifecycle"

    include Hecks::Model

    attribute :model_id
    attribute :policy_id
    attribute :reviewer_id
    attribute :outcome
    attribute :notes
    attribute :completed_at
    attribute :conditions, default: [], freeze: true
    attribute :status

    # State predicates — see lifecycle.rb for full state machine
    def open?; status == "open"; end
    def approved?; status == "approved"; end
    def rejected?; status == "rejected"; end
    def changes_requested?; status == "changes_requested"; end

    VALID_OUTCOME = ["approved", "rejected"].freeze unless defined?(VALID_OUTCOME)

    private

    def validate!
      raise ValidationError.new("model_id can't be blank", field: :model_id, rule: :presence) if model_id.nil? || (model_id.respond_to?(:empty?) && model_id.empty?)
      raise ValidationError.new("reviewer_id can't be blank", field: :reviewer_id, rule: :presence) if reviewer_id.nil? || (reviewer_id.respond_to?(:empty?) && reviewer_id.empty?)
      if outcome && !VALID_OUTCOME.include?(outcome)
        raise ValidationError, "outcome must be one of: #{VALID_OUTCOME.join(', ')}, got: #{outcome}"
      end
    end
  end
end
